domains/client/relayer/src/worker.rs

```diff
@@ -174,50 +219,32 @@ where
         + AuxStore
         + ProofProvider<Block>
         + ProvideRuntimeApi<Block>,
-    Client::Api: RelayerApi<Block, NumberFor<Block>>,
-    MP: Fn(&Arc<Client>, Block::Hash) -> Result<(), Error>,
+    MP: Fn(&Arc<Client>, (NumberFor<Block>, Block::Hash), ExtraData) -> Result<(), Error>,
     SO: SyncOracle,
-    CRM: Fn(ChainId, NumberFor<Block>) -> Result<bool, ApiError>,
+    CRM: Fn(NumberFor<Block>) -> Result<Option<ExtraData>, ApiError>,
 {
-    let chain_id = Relayer::chain_id(&client)?;
     tracing::info!(
         target: LOG_TARGET,
         "Starting relayer for chain: {:?}",
         chain_id,
     );
-    let mut chain_block_import = client.import_notification_stream();
+    let mut chain_block_finalization = client.finality_notification_stream();
 
     // from the start block, start processing all the messages assigned
-    // wait for new block import of chain,
+    // wait for new block finalization of the chain,
     // then fetch new messages in the block
     // construct proof of each message to be relayed
     // submit XDM as unsigned extrinsic.
-    while let Some(block) = chain_block_import.next().await {
+    while let Some(block) = chain_block_finalization.next().await {
         // if the client is in major sync, wait until sync is complete
         if sync_oracle.is_major_syncing() {
             tracing::debug!(target: LOG_TARGET, "Client is in major sync. Skipping...");
             continue;
         }
 
-        let relay_block_until = match block.header.number().checked_sub(&relay_confirmation_depth) {
-            None => {
-                // not enough confirmed blocks.
-                tracing::debug!(
-                    target: LOG_TARGET,
-                    "Not enough confirmed blocks for domain: {:?}. Skipping...",
-                    chain_id
-                );
-                continue;
-            }
-            Some(confirmed_block) => confirmed_block,
-        };
-
         let (number, hash) = (*block.header.number(), block.header.hash());
         let blocks_to_process: Vec<(NumberFor<Block>, Block::Hash)> =
-            Relayer::fetch_unprocessed_blocks_until(&client, chain_id, number, hash)?
-                .into_iter()
-                .filter(|(number, _)| *number <= relay_block_until)
-                .collect();
+            Relayer::fetch_unprocessed_consensus_blocks_until(&client, chain_id, number, hash)?;
```

Comment:
XDM will remain in the state when un responded, I'm not sure I clearly follow what you mean. As and when a consensus block is finalized, we relay the XDM from those blocks. 

I realized we still need a fallback mechanism for when the message was dropped sliently. What we can do here is to run a simultaneous process that re-submits the XDM. I will handle this in seperate PR and there is an issue created for this already I believe

Reply:
>XDM will remain in the state when un responded, I'm not sure I clearly follow what you mean

I mean we don't need the aux storage for tracking whether a block is processed or not.

Since the XDM is available in the state, whenever a consensus block import/finalization notification comes, we can get the XDM from the state based on the imported/finalized block, filter already relayed XDM, and relay the XDM. If an XDM is dropped silently during propagation then when the next consensus block import/finalization notification comes we will retry to relay this XDM without needing another worker. And yeah let's handle this in a seperate PR.